### DISCUSSION TITLE
Large Model with Vehicle Time Windows Constraints

### DISCUSSION THREAD

**Discussion by FreddyHunter** [google/or-tools#2929] [Routing (and legacy CP) questions]
Dear comunity,

I am using the following step-by-step to constraint location visits by each vehicle (with different time windows):

1) duplicate the location to model each time windows
2) Restrict the vehicle list of each location using routing.VehicleVar()
3) Add a disjunction to solver only have to visit one of these nodes

When I was developing it and testing with small model it was ok and running in few seconds, but when I started to test it with large model the time spent degraded a lot. For example: 100 locations and 10 vehicles it is taking around 90 seconds to create the model and simulate 10 searches, when we increase the number of vehicles to 20 keeping the same numbers of locations it goes to 17 min, and when we try it with 30 or more vehicles it "never stops". 

Without vehicle time windows constraints it run 100 searches less than 10 seconds.

Any tip or suggestion to improve it?

I am developing using latest version of OR Tools at Python.

**Comment by Mizux** [comment]
strange it impact so much the solve.
1. Could you share one instance so we can repoduce, test it on our side ?

2. If you have 10 vehicles, do you duplicate each locations 10 times **OR** do you have a pool of identical vehicles like 3 TW_0, 4 TW_1 and 3 TW_2 so you only to duplicate the node three time ?

**Comment by FreddyHunter** [comment]
Hi Mizux,

Thanks for your reply.

1) yes, I can share it for you.

2) yes, I am duplicating each location 10 times for 10 vehicles. I believe
design the pool will be a better idea, once my model has around 1000
locations and 50 vehicles. But, moving in this way how can I make the
solver select the vehicle once the pool is selected?

Thanks.

**Reply by Mizux** [reply]
I mean suppose you have pool of vehicle 0: [0, 1, 2, 3] and pool of vehicle [4, 5, 6].
you can assign a pool to a node avoiding to duplicate it 4 or 3 time more.
e.g.
```py
pool_one = [-1, 0, 1, 2, 3] # add -1 to allow node to be dropped
pool_two = [-1, 4, 5, 6]

index_node_one = manager.NodeToIndex(node_one)
routing.VehicleVar(index_node_one).SetValues(pool_one)

index_node_two = manager.NodeToIndex(node_two)
routing.VehicleVar(index_node_two).SetValues(pool_two)
```

**Comment by FreddyHunter** [comment]
Thanks Mizux, I will try that.
